### DISCUSSION TITLE
Accurately Measuring Query Performance

### DISCUSSION THREAD

**Discussion by bill3tt** [thanos-io/thanos#4674] [General]
Hello folks, let's talk about strategies for accurately measuring query performance in Thanos...

This is a topic we discussed in the [Thanos Community Office Hours](https://www.google.com/url?q=https://docs.google.com/document/d/137XnxfOT2p1NcNUq6NWZjwmtlSdA6Wyti86Pd6cyQhs/edit%23&sa=D&source=calendar&ust=1632265658837402&usg=AOvVaw0Y4w_XZhuKkHNWcNDdXQEn), but wanted to raise it properly here :)

## Context

I'm attempting to define a set of SLIs for the read path of [Red Hat's internal observability service](https://github.com/rhobs).

To do this we need to figure out a way of measuring the query path in a **reliable** and **consistent** way.

## Problem

It turns out that accurately measuring Query path latency is actually pretty tricky.

### Why can't we just measure `/query` & `/query_range` latencies?

In Thanos, like any data system, the time a read request takes to return is determined by the query itself. 

With a diverse enough set of users & queries that change over time, simply measuring server response latencies does not really tell us any consistently useful information about the performance of the system.

This feels like an unsuitable measurement for a SLI as any user could violate any SLO with a couple of big enough queries.  

### What determines query latency?

Query path performance is _primarily_ determined by the number of samples that the query touches / processes. The more samples that your query touches / processes (but does not necessarily return), the longer the query will take to return.

There are other factors that impact query latency including:
* Query shape - Requesting 1M samples from 1 series will have a different latency profile than querying 1K samples from 1K series.
* Data distribution - Requesting 'cold' samples from object storage via Store will be slower than 'hot' samples in memory in the TSDB head block.

The overarching question is: how do we measure query path performance in a **consistent** and **accurate** way?

## Solutions

It seems like any suitable measurement needs to have reference to the number of samples processed.

Off the top of my head I can think of two ways of doing this...

### Query Latency Histogram

One option would be for Thanos Query expose a histogram-like metric that allows us to express logic like: 
* What was the 99th percentile latency of queries that touched less than 1M samples?
* What was the 90th percentile latency of queries that touched less than 100M samples?

The tricky part of this is that we would kind of need a histogram of histograms? Or at least a configurable set of histograms with configurable thresholds.

This solution would be nice because we are directly measuring user queries.

### Known Load, Known Query

Another way of accurately measuring query performance over time is to write a set of synthetic data with a known cardinality to Thanos (using something like [avalanche](https://github.com/open-fresh/avalanche)) and then continuously execute a known query against this data and measure the response latencies. With this data you can then say 'the 99th percentile of query latencies that touch 1M samples is 10s for example'.

This is nice because it does not require any upstream changes, but ultimately does not _really_ measure real user query performance.

## Community

Questions for the community:
* Have you come across this problem before?
* Have you solved this problem in a different way?

cc @jmichalek132 as I know you have been thinking about this too!

**Comment by bwplotka** [comment]
![](https://media3.giphy.com/media/d8lUKXD00IXSw/giphy.gif)

**Comment by bwplotka** [comment]
On a serious note, thanks for starting this. Thanos users - if you have any input, let us know! (: 

*We can't improve something (or limit) if we don't measure it!*

We had some discussion together on Red Hat side:
* Having a probe the writes 100k series with 1000 samples each (or something) and retrieving it and measuring could be a good starting point. (How often though?)
* Ideally we can have something on request like samples or series touched. This ideally is transformed to certain `throughput` metric which is in form of  bytes or samples + series  per seconds

**Comment by dofinn** [comment]
+1 for this - could potentially be utilised further to educate querier?
